### DISCUSSION TITLE
Is there a way to publish a library only devDependency

### DISCUSSION THREAD

**Discussion by goatandsheep** [npm/feedback#1072] [Registry] [answered]
I am a maintainer for a library that only works when installed as a devDependency. I get comments every day of people having various errors from installing it accidentally as a regular dependency. Is there a way to either:

1. make it so `npm i <library>` defaults to devDependency in the project from the `package.json` of my library
2. make it so `npm i <library>` can only be installed as a devDependency in the project from the `package.json` of my library
3. make 2 separate versions of the library so if it is installed incorrectly, the wrong code runs and returns an error
4. determine that the library has been installed as a regular dependency in code such that I can fire an error

Or other options. I'm open to suggestion.

**Comment by ljharb** [comment]
How could a library only work when installed as a dev dep?

There do not exist any packages, as far as I’m aware, that are always only one kind of dependency.

**Reply by goatandsheep** [reply]
Libraries that are meant for build-time only and are not part of the project. I maintain https://github.com/goatandsheep/react-native-dotenv , which is a babel plugin most commonly used in react native projects. The babel plugin relies on fs, path, and os to read the dotenv files, whose values are written into the babel cache. When installed as a production dependency, the code fails as it attempts to compile the native binaries including fs, path, and os, causing major instability.

**Reply by ljharb** [reply]
Right, but a babel plugin might be used by another babel plugin, in which case it would be a runtime dependency of that plugin.

_Nothing_ is always only one kind of dep, including babel plugins. That means that there's no real alternative besides educating your users with good docs and error messages.

**Reply by goatandsheep** [reply, marked as answer]
@ljharb I agree. And this is why I didn't even consider doing something like this before. However, I'm noticing an increasing number of users installing it as a production dependency despite multiple warnings in the docs. Perhaps I just need to add it to the issue template.
